### PR TITLE
[ISSUE #1013] Fix: select one message queue of different broker when retry to send

### LINKED ISSUE #1013: It will stop to retry when use SyncSend to send in some case
**BUG REPORT**  

Currently in SyncSend, retries are only worked when there is a network error, but not for common errors such as Broker busy
```golang
res, _err := p.client.InvokeSync(ctx, addr, p.buildSendRequest(mq, msg), 3*time.Second)
if _err != nil {
	err = _err
	continue
}

return p.client.ProcessSendResponse(mq.BrokerName, res, resp, msg)
```
In [java SDK](https://github.com/apache/rocketmq/blob/dcf7ccdd9ba92c22a44cde4c1df36c4d1ef6a172/client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java#LL615C17-L615C17), It will retry for `broker_busy`、`servicce not available` and so on. And It will select a different broker to retry. 

**This is very helpful in a multi-Broker cluster, so I think it is necessary to fix this problem.**

### PR DESCRIPTION
**Pull request by cserwen**
## What is the purpose of the change

fix #1013 

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when a cross-module dependency exists.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Review by francisoliverlee — CHANGES_REQUESTED** [review]
建议添加一些描述， 说明bug出现的场景和修复方法。方便大家review和理解代码逻辑

**Comment by cserwen** [comment]
> 建议添加一些描述， 说明bug出现的场景和修复方法。方便大家review和理解代码逻辑

@francisoliverlee  Hi, please see issue #1014.

File: producer/selector_test.go [hunk 1]
```diff
@@ -26,7 +26,7 @@ import (
 )
 
 func TestRoundRobin(t *testing.T) {
-	queues := make([]*primitive.MessageQueue, 10)
```
Comment (ferrirW) [inline producer/selector_test.go:29]: why 10-> 0 @cserwen

**Review by ferrirW — APPROVED** [review]

File: producer/selector_test.go (hunk 1 above)
Comment (cserwen) [reply producer/selector_test.go:29]: make() will init the slice 
image

**Review by ferrirW — APPROVED** [review]

**Review by humkum — APPROVED** [review]

**Comment by twz915** [comment]
@cserwen 

```
		mq := p.selectMessageQueue(msg, lastBrokerName)
```
this is a bug, which should be 
```
		mq = p.selectMessageQueue(msg, lastBrokerName)
```

**Pull request merged**